File: 01-coding/golang/health_checker.go
package main

/*
HTTP Health Checker - Go Edition

Demonstrates:
- HTTP client usage
- Concurrency with goroutines
- Channels for communication
- Context for timeouts
*/

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"time"
)

// HealthCheck represents a health check result
type HealthCheck struct {
	URL          string
	StatusCode   int
	ResponseTime time.Duration
	Healthy      bool
	Error        error
}

// CheckHealth performs a health check on a URL
func CheckHealth(ctx context.Context, url string) HealthCheck {
	start := time.Now()
	result := HealthCheck{
		URL:     url,
		Healthy: false,
	}

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		result.Error = err
		return result
	}

	client := &http.Client{
		Timeout: 5 * time.Second,
	}

	resp, err := client.Do(req)
	if err != nil {
		result.Error = err
		return result
	}
	defer resp.Body.Close()

	result.StatusCode = resp.StatusCode
	result.ResponseTime = time.Since(start)
	result.Healthy = resp.StatusCode >= 200 && resp.StatusCode < 300

	return result
}

// CheckMultipleServices checks multiple services concurrently
func CheckMultipleServices(urls []string) []HealthCheck {
	results := make([]HealthCheck, len(urls))
	resultChan := make(chan HealthCheck, len(urls))

	// Start health checks concurrently
	for _, url := range urls {
		go func(u string) {
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			defer cancel()
			resultChan <- CheckHealth(ctx, u)
		}(url)
	}

	// Collect results
	for i := 0; i < len(urls); i++ {
		results[i] = <-resultChan
	}

	return results
}

func main() {
	urls := []string{
		"https://www.google.com",
		"https://www.github.com",
		"https://httpbin.org/status/200",
	}

	if len(os.Args) > 1 {
		urls = os.Args[1:]
	}

	fmt.Println("🏥 Health Check Tool")
	fmt.Println("=" * 50)
	fmt.Printf("\nChecking %d service(s)...\n\n", len(urls))

	results := CheckMultipleServices(urls)

	// Display results
	healthyCount := 0
	for _, result := range results {
		status := "❌ UNHEALTHY"
		if result.Healthy {
			status = "✅ HEALTHY"
			healthyCount++
		}

		fmt.Printf("%s %s\n", status, result.URL)
		if result.Error != nil {
			fmt.Printf("  Error: %v\n", result.Error)
		} else {
			fmt.Printf("  Status: %d, Response Time: %v\n", result.StatusCode, result.ResponseTime)
		}
		fmt.Println()
	}

	fmt.Println("=" * 50)
	fmt.Printf("Summary: %d/%d services healthy\n", healthyCount, len(urls))

	if healthyCount < len(urls) {
		os.Exit(1)
	}
}

File: 01-coding/python/health_checker.py
#!/usr/bin/env python3
"""
Service Health Checker

This script demonstrates:
- HTTP health checks
- Retry logic with exponential backoff
- Structured logging
- Exit codes for monitoring integration
"""

import sys
import time
import logging
import argparse
from typing import Dict, Optional

try:
    import requests
except ImportError:
    print("Error: requests library not installed. Run: pip install requests")
    sys.exit(1)

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class HealthChecker:
    """Service health checker with retry logic."""
    
    def __init__(self, url: str, timeout: int = 5, max_retries: int = 3):
        self.url = url
        self.timeout = timeout
        self.max_retries = max_retries
    
    def check_health(self) -> Dict[str, any]:
        """
        Check service health with retry logic.
        
        Returns:
            dict: Health check results including status, response time, etc.
        """
        for attempt in range(1, self.max_retries + 1):
            try:
                logger.info(f"Health check attempt {attempt}/{self.max_retries} for {self.url}")
                
                start_time = time.time()
                response = requests.get(self.url, timeout=self.timeout)
                response_time = time.time() - start_time
                
                result = {
                    'url': self.url,
                    'status_code': response.status_code,
                    'response_time': round(response_time, 3),
                    'healthy': 200 <= response.status_code < 300,
                    'attempt': attempt
                }
                
                if result['healthy']:
                    logger.info(f"✓ Service healthy - Status: {result['status_code']}, "
                              f"Response time: {result['response_time']}s")
                    return result
                else:
                    logger.warning(f"✗ Service unhealthy - Status: {result['status_code']}")
                    
            except requests.exceptions.Timeout:
                logger.error(f"Timeout on attempt {attempt}")
            except requests.exceptions.ConnectionError:
                logger.error(f"Connection error on attempt {attempt}")
            except Exception as e:
                logger.error(f"Unexpected error on attempt {attempt}: {e}")
            
            # Exponential backoff
            if attempt < self.max_retries:
                wait_time = 2 ** attempt
                logger.info(f"Waiting {wait_time}s before retry...")
                time.sleep(wait_time)
        
        # All retries failed
        return {
            'url': self.url,
            'healthy': False,
            'error': 'All retries exhausted'
        }


def main():
    parser = argparse.ArgumentParser(description='Service Health Checker')
    parser.add_argument('url', help='URL to check')
    parser.add_argument('--timeout', type=int, default=5, help='Request timeout in seconds')
    parser.add_argument('--retries', type=int, default=3, help='Number of retry attempts')
    
    args = parser.parse_args()
    
    checker = HealthChecker(args.url, args.timeout, args.retries)
    result = checker.check_health()
    
    # Exit with appropriate code for monitoring systems
    if result['healthy']:
        print(f"\n✓ Service is HEALTHY")
        print(f"  Response time: {result.get('response_time', 'N/A')}s")
        return 0
    else:
        print(f"\n✗ Service is UNHEALTHY")
        print(f"  Error: {result.get('error', 'Unknown error')}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
